contracts/0.8.25/vaults/VaultHub.sol

```diff
@@ -519,202 +612,487 @@ contract VaultHub is PausableUntilWithRoles {
+        // Impossible to rebalance a vault with deficit
+        if (liabilityStETH >= totalValue_) {
+            // return MAX_UINT_256
+            return type(uint256).max;
         }
+
+        // Solve the equation for X:
+
+        // LS - liabilityStETH, TV - totalValue, MR - maxMintableRatio, 100 - TOTAL_BASIS_POINTS, RR - reserveRatio
+
+        // X - amount of ETH that should be withdrawn (TV - X) and used to repay the debt (LS - X)
+        // to reduce the LS/TV ratio back to MR
+
+        // (LS - X) / (TV - X) = MR / 100
+        // (LS - X) * 100 = (TV - X) * MR
+        // LS * 100 - X * 100 = TV * MR - X * MR
+        // X * MR - X * 100 = TV * MR - LS * 100
+        // X * (MR - 100) = TV * MR - LS * 100
+        // X = (TV * MR - LS * 100) / (MR - 100)
+        // X = (LS * 100 - TV * MR) / (100 - MR)
+        // RR = 100 - MR
+        // X = (LS * 100 - TV * MR) / RR
+
+        return (liabilityStETH * TOTAL_BASIS_POINTS - totalValue_ * maxMintableRatio) / reserveRatioBP;
     }
 
-    function mintVaultsTreasuryFeeShares(uint256 _amountOfShares) external {
-        if (msg.sender != LIDO_LOCATOR.accounting()) revert NotAuthorized("mintVaultsTreasuryFeeShares", msg.sender);
-        LIDO.mintExternalShares(LIDO_LOCATOR.treasury(), _amountOfShares);
+    /// @dev check if the share limit is within the upper bound set by MAX_RELATIVE_SHARE_LIMIT_BP
+    function _maxSaneShareLimit() internal view returns (uint256) {
+        return (LIDO.getTotalShares() * MAX_RELATIVE_SHARE_LIMIT_BP) / TOTAL_BASIS_POINTS;
+    }
+
+    function _unlocked(VaultRecord storage _record) internal view returns (uint256) {
+        uint256 totalValue_ = _totalValue(_record);
+        uint256 locked_ = _record.locked;
+
+        if (locked_ > totalValue_) return 0;
+
+        return totalValue_ - locked_;
     }
 
-    function _vaultAuth(address _vault, string memory _operation) internal view {
-        if (msg.sender != OwnableUpgradeable(_vault).owner()) revert NotAuthorized(_operation, msg.sender);
+    function _totalValue(VaultRecord storage _record) internal view returns (uint256) {
+        Report memory report = _record.report;
+        return uint256(int256(uint256(report.totalValue)) + _record.inOutDelta - report.inOutDelta);
     }
 
-    function _connectedSocket(address _vault) internal view returns (VaultSocket storage) {
-        VaultHubStorage storage $ = _getVaultHubStorage();
-        uint256 index = $.vaultIndex[_vault];
-        if (index == 0 || $.sockets[index].pendingDisconnect) revert NotConnectedToHub(_vault);
-        return $.sockets[index];
+    function _isReportFresh(VaultRecord storage _record) internal view returns (bool) {
+        uint256 latestReportTimestamp = LazyOracle(LIDO_LOCATOR.lazyOracle()).latestReportTimestamp();
+        return
+            // check if AccountingOracle brought fresh report, but allow freshly connected vaults
+            latestReportTimestamp <= _record.reportTimestamp &&
```

Comment:
:thinking: probably we can set the default connection timestamp by retrieving it from LazyOracle and move from `<=` to `==` check here